### PR TITLE
Add support for concurrent writes to DataObjectLockFree

### PR DESCRIPTION
**Pull request by meyerj**
Required to fully support `PerInputPort` and `Shared` buffer policies on lock-free data connections. In order to avoid that multiple writer threads start to assign to the same storage elements pointed to by `write_ptr`, a new atomic counter has been added that allows the threads to detect the situation and start over again. Additionally, the `Get()` method has been updated to apply compare-and-swap on the `FlowStatus` stored in the buffer, so that at most one reader thread returns `NewData` for each sample.

The refactoring of the `RTT::os::SharedMutex` class is only vaguely related, but it removes the mutex that has been used to protect the state information which could be problematic and cause priority inversion issues. Non-exclusive locks on shared mutexes are used extensively in the new dataflow implementation introduced in https://github.com/orocos-toolchain/rtt/pull/114 to protect the channel pipeline from concurrent connects and disconnects, but in most cases they are only ever locked by a single thread. On the other hand a comparison with the previous implementation did not show any performance improvements using the `gnulinux` target, probably because pthread mutexes are already optimized for the case that there is no congestion.

See http://preshing.com/20120612/an-introduction-to-lock-free-programming/ for a good introduction in the topic of lock-free programming and http://preshing.com/20150316/semaphores-are-surprisingly-versatile/#a-lightweight-semaphore-with-partial-spinning for the slightly adapted idea of implementing the lock-free read-write-locks using semaphores.

Once this patch has been reviewed and tested I will port the patches to upstream RTT.

@psoetens
@smits
@snrkiwi

### CODE REVIEW

File: rtt/base/DataObjectLockFree.hpp [hunk 1]
```diff
@@ -249,38 +254,49 @@ namespace RTT
...
 
-            // writeout in any case
-            PtrType writing = write_ptr;
+            PtrType writing;
+            do {
+                writing = write_ptr;            // copy buffer location
+                oro_atomic_inc(&writing->write_counter); // lock buffer, no more writes
+                // XXX smp_mb
+                if ( writing != write_ptr )     // if write_ptr changed,
+                    oro_atomic_dec(&writing->write_counter); // better to start over.
```
Comment (meyerj) [inline rtt/base/DataObjectLockFree.hpp]: Eventually we could simply abort here instead of trying again. Another thread was faster and modified the `write_ptr` already since we made a copy in line 265. At the moment we prefer the sample written by low-priority threads because they will overwrite previous samples after they are unblocked. If `Set()` would abort as soon as it detects attempts to write concurrently, high-priority threads would have a higher chance to "push" their sample.

File: rtt/base/BufferLockFree.hpp [hunk 2]
```diff
@@ -180,8 +180,8 @@ namespace RTT
 
         bool Push( param_t item)
         {
-            if ( capacity() == (size_type)bufs->size() ) {
-                if (!mcircular)
+            if (!mcircular) {
+                if ( capacity() == (size_type)bufs->size() )
```
Comment (kevindm) [inline rtt/base/BufferLockFree.hpp]: As a matter of taste, the following seems more concise and cleaner.
```
if (!mCircular && capacity() == (size_type)bufs->size()) {
  return false;
}
```

File: rtt/os/oro_i386/oro_arch.h [hunk 3]
```diff
@@ -64,6 +64,17 @@ static __inline__ void oro_atomic_add( oro_atomic_t *v, int i)
...
 
+static __inline__ int oro_atomic_add_return( oro_atomic_t *v, int i)
+{
+  /* Modern 486+ processor */
+	int __i = i;
+	__asm__ __volatile__(
+		ORO_LOCK "xaddl %0, %1"
+		: "+r" (i), "+m" (v->counter)
+		: : "memory");
+	return i + __i;
```
Comment (kevindm) [inline rtt/os/oro_i386/oro_arch.h:75]: Variable naming could be improved. `i` and `__i`  are dubiously close together. My first impression was: 'does this return 2*i'?
How about replacing `int __i = i;` by `register counter = i;` and use counter in the asm instead of i?  I'm not sure about the code as is not incurring an additional store operation after the assembly (it may get optimized-out though).

File: rtt/os/oro_noasm/oro_arch.h [hunk 4]
```diff
@@ -73,11 +73,25 @@ static __inline__ void oro_atomic_add(oro_atomic_t *a_int, int n )
   rtos_mutex_lock(&((a_int)->m)); (a_int)->cnter += n; rtos_mutex_unlock(&((a_int)->m));
 }
 
+static __inline__ int oro_atomic_add_return(oro_atomic_t *a_int, int n )
+{
+  int new_value;
+  rtos_mutex_lock(&((a_int)->m)); new_value = ((a_int)->cnter += n); rtos_mutex_unlock(&((a_int)->m));
```
Comment (kevindm) [inline rtt/os/oro_noasm/oro_arch.h:79]: What is gained from performing this operation on a single line?

File: rtt/os/oro_i386/oro_arch.h [hunk 5]
```diff
@@ -64,6 +64,17 @@ static __inline__ void oro_atomic_add( oro_atomic_t *v, int i)
 		:"ir" (i), "m" (v->counter));
 }
 
+static __inline__ int oro_atomic_add_return( oro_atomic_t *v, int i)
+{
+  /* Modern 486+ processor */
```
Comment (kevindm) [inline rtt/os/oro_i386/oro_arch.h:69]: This comment seems to clash with the file name. i386 is not a modern 486+ processor.

File: tests/ports_test.cpp [hunk 6]
```diff
@@ -922,10 +922,10 @@ BOOST_AUTO_TEST_CASE(testEventPortSignalling)
     OutputPort<double> wp1("Write");
     InputPort<double>  rp1("Read");
 
-    BOOST_ASSERT(tce->configure());
-    BOOST_ASSERT(tce->isConfigured());
+    BOOST_REQUIRE(tce->configure());
```
Comment (kevindm) [inline tests/ports_test.cpp:925]: Out of curiosity: Why this change?

File: rtt/os/CAS.hpp [hunk 7]
```diff
@@ -51,10 +51,17 @@ namespace RTT
...
-    bool CAS( volatile T* addr, const V& expected, const W& value) {
+    inline bool CAS( volatile T* addr, const V& expected, const W& value) {
         return expected == oro_cmpxchg(addr, expected, value);
     }
 
+    /**
+     * Overload of Compare And Swap for oro_atomic_t.
+     */
+    inline bool CAS( volatile oro_atomic_t* addr, const int expected, const int value) {
+        return expected == oro_cmpxchg(&oro_atomic_read(addr), expected, value);
```
Comment (kevindm) [inline rtt/os/CAS.hpp]: Ugly, this looks like taking the address of a return value. Probably necessary because `oro_atomic_t` seems to be architecture dependent?

Perhaps this belongs in architecture instead?

File: rtt/base/DataObjectLockFree.hpp [hunk 8]
```diff
@@ -249,38 +254,55 @@ namespace RTT
...
-            while ( oro_atomic_read( &write_ptr->next->counter ) != 0 || write_ptr->next == read_ptr )
+            while ( oro_atomic_read( &write_ptr->next->read_counter ) != 0 ||
+                    write_ptr->next == read_ptr )
                 {
                     write_ptr = write_ptr->next;
-                    if (write_ptr == writing)
-                        return false; // nothing found, to many readers !
+                    if (write_ptr == writing) {
+                        oro_atomic_dec(&writing->write_lock);
+                        return false; // nothing found, too many writers or readers !
```
Comment (kevindm) [inline rtt/base/DataObjectLockFree.hpp]: Too many writers is impossible. There can only ever be 1 thread in this section of code.

File: rtt/base/DataObjectLockFree.hpp [hunk 9]
```diff
@@ -249,38 +254,55 @@ namespace RTT
...
+            // have exclusive access
+
+            // copy sample
             writing->data = push;
             writing->status = NewData;
 
             // if next field is occupied (by read_ptr or counter),
             // go to next and check again...
-            while ( oro_atomic_read( &write_ptr->next->counter ) != 0 || write_ptr->next == read_ptr )
+            while ( oro_atomic_read( &write_ptr->next->read_counter ) != 0 ||
```
Comment (kevindm) [inline rtt/base/DataObjectLockFree.hpp]: As discussed, this loop should probably not directly manipulate the write_ptr, to prevent preemption in the middle by a writer that would subsequently write to a buffer that is (potentially) being read.

**Review by kevindm — COMMENTED** [review]
I had some additional comments that were unrelated to the correctness discussion we had offline.

File: tests/ports_test.cpp (hunk 6 above)
Comment (meyerj) [reply tests/ports_test.cpp:925]: It is explained in the commit message, https://github.com/meyerj/rtt/pull/9/commits/c892abbb3085b33118e0283f38329e2db6ce6d3d.

`BOOST_ASSERT()` is not a macro provided by the [unit testing framework](http://www.boost.org/doc/libs/1_65_1/libs/test/doc/html/boost_test/utf_reference/testing_tool_ref.html) and is only effective in debug builds (without `-NDEBUG`). Should have been `BOOST_REQUIRE()`, which is equivalent to gtest's `ASSERT_TRUE()`.

**Comment by meyerj** [comment]
With https://github.com/meyerj/rtt/pull/9/commits/784c16100ed512dd9161e290d3cfd6e89b2f2894 the observable behavior of `DataObjectLockFree<T>` in case of multiple concurrent writes became similar to a critical section protected by a `MutexTryLock`, where the first thread that starts to write always wins, even it it has a lower priority, and all other writers drop out early and fail. Therefore samples being written by low-priority writers have a higher chance to be "seen" by readers and writes by high-priority threads are more likely to fail. While this might be unexpected on first sight, it avoids problems while updating the read pointer safely and should not make a difference for most applications. Writing to a lock-free data object is even wait-free.

The alternative would be that two writer threads can start to assign to different buffer elements concurrently and the first thread that updates the read pointer successfully "wins", while the other has to detect that situation and simply skip updating the read pointer again. Like this, samples being written by high-priority writers would have a higher chance to be "seen" by readers.

File: rtt/base/DataObjectLockFree.hpp (hunk 9 above)
Comment (meyerj) [reply rtt/base/DataObjectLockFree.hpp]: If all other writers drop out early, like it is now, there is no problem because it is guaranteed that no other thread enters that section concurrently.

File: rtt/os/CAS.hpp (hunk 7 above)
Comment (meyerj) [reply rtt/os/CAS.hpp]: I agree that it looks ugly, but it's necessary to allow CAS for atomic integers. `oro_atomic_read()` is a macro that resolves to something like `(addr->cnter)` and it is save to take the address of the return value. If the definition of the macro would change for some of the supported architectures, this has to be reevaluated.

File: rtt/os/oro_i386/oro_arch.h (hunk 5 above)
Comment (meyerj) [reply rtt/os/oro_i386/oro_arch.h:69]: But 486+ processors still have i386 architecture.

File: rtt/os/oro_i386/oro_arch.h (hunk 3 above)
Comment (meyerj) [reply rtt/os/oro_i386/oro_arch.h:75]: You should not spend time on commenting on `oro_arch.h` implementations. This code has been basically copied from the Linux kernel: https://elixir.free-electrons.com/linux/v2.6.39/source/arch/x86/include/asm/atomic.h#L173, like the other functions in those files.

File: rtt/os/oro_noasm/oro_arch.h (hunk 4 above)
Comment (meyerj) [reply rtt/os/oro_noasm/oro_arch.h:79]: Nothing but consistent code style.

File: rtt/base/DataObjectLockFree.hpp (hunk 8 above)
Comment (meyerj) [reply rtt/base/DataObjectLockFree.hpp]: This was a leftover from the first implementation that would have allowed for concurrent writes...

**Comment by meyerj** [comment]
I will open another PR against upstream toolchain-2.9 in order to involve the community.

File: rtt/os/oro_i386/oro_arch.h (hunk 3 above)
Comment (kevindm) [reply rtt/os/oro_i386/oro_arch.h:75]: The copyright notice should mention this, I think.

File: rtt/os/oro_i386/oro_arch.h (hunk 5 above)
Comment (kevindm) [reply rtt/os/oro_i386/oro_arch.h:69]: In a sense. Does the assembly operation exist in i386 or only 486+? Not that important, it is just a comment.

File: rtt/base/DataObjectLockFree.hpp (hunk 9 above)
Comment (kevindm) [reply rtt/base/DataObjectLockFree.hpp]: When you get preempted in the middle of this loop, stuff could go wrong, no?

File: rtt/os/oro_i386/oro_arch.h (hunk 5 above)
Comment (meyerj) [reply rtt/os/oro_i386/oro_arch.h:69]: Apparently `xaddl` was not yet supported or buggy before 486 and the kernel developers decided to fall back to a generic "disable global interrupts" implementation under certain conditions, but I decided to omit this code path and assume that we will always run on a "modern 486+ processor" and did not spend time to figure out how we could do something equivalent in user-space.

File: rtt/os/oro_i386/oro_arch.h (hunk 3 above)
Comment (meyerj) [reply rtt/os/oro_i386/oro_arch.h:75]: Copyright notices are a mess anyway and worth a major cleanup. But it is hard to reproduce the full history and contributors of each file. RTT has been moved from SVN to git, files have been moved within the repo and not all sources have been properly tracked. The oldest commit still present is https://github.com/meyerj/rtt/commit/6216aba84460c785d545d7b570f678148b9958ed, without an extra copyright notice.

File: rtt/os/oro_i386/oro_arch.h (hunk 5 above)
Comment (kevindm) [reply rtt/os/oro_i386/oro_arch.h:69]: You should probably mention this in the comment.

File: rtt/base/DataObjectLockFree.hpp (hunk 9 above)
Comment (meyerj) [reply rtt/base/DataObjectLockFree.hpp]: I don't see why. If only one thread is executing that finite loop, and no other thread ever waits for its completion, there is no problem. Or do I miss something?
